hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/common/HoodieSparkEngineContext.java

```diff
@@ -148,6 +148,7 @@ public <I, K, V> Stream<ImmutablePair<K, V>> mapPartitionsToPairAndReduceByKey(
                 .map(e -> new Tuple2<>(e.getKey(), e.getValue())).iterator()
         )
         .reduceByKey(reduceFunc::apply)
+        .coalesce(parallelism)
```

Comment:
Should we also revisit other API implementations in `HoodieSparkEngineContext` for applying parallelism for the shuffling?

Reply:
HUDI-9111 to follow up.